scheduler/src/cook/kubernetes/compute_cluster.clj

```diff
@@ -370,125 +374,134 @@
-    (try
-      (assert (cc/autoscaling? this pool-name)
-              (str "In " name " compute cluster, request to autoscale despite invalid / missing config"))
-      (let [timer-context-autoscale (timers/start (metrics/timer "cc-synthetic-pod-autoscale" name))
-            outstanding-synthetic-pods (->> (get-pods-in-pool this pool-name)
-                                            (add-starting-pods this)
-                                            (filter synthetic-pod->job-uuid))
-            num-synthetic-pods (count outstanding-synthetic-pods)
-            total-pods (-> @all-pods-atom keys count)
-            total-nodes (-> @current-nodes-atom keys count)
-            {:keys [image user command max-pods-outstanding max-total-pods max-total-nodes]
-             :or {command "exit 0" max-total-pods 32000 max-total-nodes 1000}} synthetic-pods-config]
-        (log/info "In" name "compute cluster there are" total-pods "pods and" total-nodes
-                  "nodes of a max of" max-total-pods "pods and" max-total-nodes "nodes")
-        (log/info "In" name "compute cluster, for pool" pool-name "there are" num-synthetic-pods
-                  "outstanding synthetic pod(s), and a max of" max-pods-outstanding "are allowed")
-        (let [max-launchable (min (- max-pods-outstanding num-synthetic-pods)
-                                  (- max-total-nodes total-nodes)
-                                  (- max-total-pods total-pods))]
-          (if (not (pos? max-launchable))
-            (log/info "In" name "compute cluster, cannot launch more synthetic pods")
-            (let [using-pools? (config/default-pool)
-                  synthetic-task-pool-name (when using-pools? pool-name)
-                  new-jobs (remove (fn [{:keys [job/uuid]}]
-                                     (some #(= (str uuid) (synthetic-pod->job-uuid %))
-                                           outstanding-synthetic-pods))
-                                   jobs)
-                  user-from-synthetic-pods-config user
-                  task-metadata-seq
-                  (->> new-jobs
-                       (map (fn [{:keys [job/user job/uuid job/environment] :as job}]
-                              (let [pool-specific-resources
-                                    ((adjust-job-resources-for-pool-fn pool-name) job (tools/job-ent->resources job))]
-                                {:command {:user (or user-from-synthetic-pods-config user)
-                                           :value command}
-                                 :container {:docker {:image image}}
-                                 ; We need to *not* prevent the cluster autoscaler from
-                                 ; removing a node just because it's running synthetic pods
-                                 :pod-annotations {api/k8s-safe-to-evict-annotation "true"}
-                                 ; Job constraints need to be expressed on synthetic
+    (let [state @state-atom]
```

Comment:
FYI: Continung a conversation that got marked as outdated. 

@nsinkov  said  "I see, but we have no reason to expect the calls to autoscaling? and autoscaling! being atomic wrt changes to the state-atom"

Yes. There is a race here. However, in the non-race case, we will never call autoscale! if autoscaling? is false, which means this log/warn will never be printed in the non-race case. 

That doesn't seem intentional. thus, I suggest either copying it or moving it to the autoscale? line above, so that we'll get a log line if autoscalign is enabled for a non-running cluster.

Reply:
I don't know what you mean by the "non-race case"

I'm also happy not having a check for state in autoscaling?. That function just reflects the kind of cluster, not the state.

Reply:
wait, I misread. Please ignore the above comment

I think the code is correct and the warn below should be info